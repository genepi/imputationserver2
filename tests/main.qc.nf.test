nextflow_pipeline {

	def ONLY_IN_INPUT = 78;

    name "Pipeline Tests for Quality Control"
    script "main.nf"

    test("Should run QC-only") {

        when {
            params {
                project  = "test-job"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/*.vcf.gz"
                allele_frequency_population = "eur"
                mode = "qc-only"
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
            }   
        }

        then {
            assert workflow.success

            assert !file("${outputDir}/chr_20.zip").exists()
            assert !file("${outputDir}/chr20.dose.vcf.gz").exists()

            def log = file("${outputDir}/qc_report.txt")
            
            assert snapshot(
                log.text,
                file("${outputDir}/statistics/snps-typed-only.txt").text,
                file("${outputDir}/statistics/snps-excluded.txt").text
            ).match()
       
        }

    }

    test("Should run QC-only including VCF 4.3 and indels") {

        when {
            params {
                project  = "test-job"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/with_indels/*.vcf.gz"
                allele_frequency_population = "eur"
                mode = "qc-only"
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2-hg38/cloudgene.yaml"
                output = "${outputDir}"
            }   
        }

        then {
            assert workflow.success
        }

    }

    test("Should fail QC-only but publish files") {

        when {
            params {
                project  = "test-job"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/*.vcf.gz"
                allele_frequency_population = "eur"
                mode = "qc-only"
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene_with_qcfilters.yaml"
                output = "${outputDir}"
            }   
        }

        then {
            assert workflow.failed

            assert file("${outputDir}/statistics/snps-typed-only.txt").exists()
            
            def log = file("${outputDir}/qc_report.txt")

            assert snapshot(
                log.text
            ).match()
       
        }

    }

    test("Should run QC-only without population") {

        when {
            params {
                project  = "test-job"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/*.vcf.gz"
                mode = "qc-only"
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
            }   
        }

        then {
            assert workflow.success

            assert !file("${outputDir}/chr_20.zip").exists()
            assert !file("${outputDir}/chr20.dose.vcf.gz").exists()

            def log = file("${outputDir}/qc_report.txt")
            
            assert snapshot(
                log.text,
                file("${outputDir}/statistics/snps-typed-only.txt").text,
                file("${outputDir}/statistics/snps-excluded.txt").text
            ).match()
       
        }

    }

    test("Should write typed only sites") {

        when {
            params {
                project  = "testWriteTypedSitesOnly"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-unphased/*.vcf.gz"
                allele_frequency_population = "eur"
                password = PASSWORD
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
                mode = "qc-only"
                phasing = [
                    engine: "eagle"                
                ]
                imputation = [
                    min_r2: 0.5                
                ]                      
            }
        }

        then {
            assert workflow.success

            def log = file("${outputDir}/qc_report.txt")

            assert snapshot(
                log.text,
                file("${outputDir}/statistics/snps-typed-only.txt").text,
            ).match()
            
            // file starts with #, which is ignored by Tablesaw (also not used as header)
            with(path("${outputDir}/statistics/snps-typed-only.txt").csv(header: true)) {
                assert rowCount == ONLY_IN_INPUT
            }
  
        }

    }    
    
}
