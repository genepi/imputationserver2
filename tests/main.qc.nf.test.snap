{
    "Should write typed only sites": {
        "content": [
            "::log:: Reference Panel Ranges: genome-wide\n::message:: Calculating QC Statistics\n::group type=message::\n<b>Statistics:</b>\nAlternative allele frequency > 0.5 sites: 2,296\nReference Overlap: 99.00 %\nMatch: 7,735\nAllele switch: 0\nStrand flip: 0\nStrand flip and allele switch: 0\nA/T, C/G genotypes: 0\n<b>Filtered sites:</b>\nFilter flag set: 0\nInvalid alleles: 0\nMultiallelic sites: 0\nDuplicated sites: 0\nNonSNP sites: 0\nMonomorphic sites: 11\nAllele mismatch: 0\nSNPs call rate < 90%: 0\n::endgroup::\n::group type=warning::\nExcluded sites in total: 11\nRemaining sites in total: 7,735\nSee snps-excluded.txt for details\nTyped only sites: 78\nSee typed-only.txt for details\n\n::endgroup::\n",
            "ID\tCHROM\tPOS\tREF\tALT\n.\t20\t161502\tC\tT\nSNP20-723376\t20\t775376\tT\tG\nrs535315\t20\t775842\tC\tT\n.\t20\t1438230\tC\tT\nrs6753\t20\t2638579\tT\tC\nrs1127354\t20\t3193842\tC\tA\nrs622099\t20\t3580168\tG\tA\nrs6052105\t20\t3811433\tT\tC\nrs6116189\t20\t4063862\tA\tG\nrs6139486\t20\t4578080\tG\tA\nSNP20-7489572\t20\t7541572\tA\tG\nrs6140334\t20\t7581461\tT\tG\n.\t20\t7928922\tA\tG\n.\t20\t8364799\tG\tT\n.\t20\t10243251\tT\tC\nrs1926936\t20\t11005238\tA\tC\nrs2103899\t20\t12173825\tT\tC\nrs6078983\t20\t13290738\tT\tC\n.\t20\t13582476\tA\tC\nrs4441543\t20\t13650883\tT\tG\nrs6042449\t20\t13929943\tC\tT\nrs2263684\t20\t13998098\tG\tA\nrs4490256\t20\t14794361\tA\tG\n.\t20\t14921606\tG\tA\nrs6075004\t20\t16066764\tG\tA\n.\t20\t16070941\tC\tT\n.\t20\t16116624\tG\tA\n.\t20\t16120763\tA\tG\nSNP20-17544155\t20\t17596155\tC\tA\nrs11960\t20\t17600357\tG\tA\nrs1073051\t20\t17845755\tT\tC\n.\t20\t18284404\tC\tT\n.\t20\t18389203\tT\tC\n.\t20\t19025636\tC\tT\n.\t20\t20130511\tT\tC\n.\t20\t20516020\tT\tC\n.\t20\t21628586\tC\tT\n.\t20\t21863397\tA\tG\n.\t20\t21939656\tC\tT\nrs6036488\t20\t23682810\tA\tG\n.\t20\t23725019\tA\tG\n.\t20\t24261606\tT\tC\n.\t20\t24691928\tT\tG\n.\t20\t24955981\tT\tC\n.\t20\t25136824\tA\tG\nrs6132909\t20\t25948327\tA\tG\n.\t20\t26307797\tT\tC\nrs1133358\t20\t31590686\tC\tA\nSNP20-35106470\t20\t35673056\tC\tT\n.\t20\t39048873\tA\tC\nrs2228246\t20\t39792063\tA\tG\nrs6072382\t20\t40093956\tA\tC\nrs3092716\t20\t40387411\tT\tC\nSNP20-40139709\t20\t40706295\tT\tC\nrs2425656\t20\t43506247\tG\tA\nrs1116100\t20\t43668473\tG\tA\n.\t20\t44011546\tC\tT\n.\t20\t46332216\tT\tC\nrs4810699\t20\t46435339\tT\tC\nrs4809715\t20\t47263751\tT\tC\n.\t20\t47479776\tC\tT\nrs522962\t20\t48143362\tT\tC\n.\t20\t48403371\tT\tC\nrs6067931\t20\t50395283\tT\tC\n.\t20\t50475418\tG\tA\n.\t20\t53251910\tG\tA\n.\t20\t53515863\tC\tA\n.\t20\t54264333\tC\tT\nrs6064338\t20\t54692704\tA\tG\nrs1047972\t20\t54961463\tT\tC\nrs4811958\t20\t56637477\tA\tG\n.\t20\t57304273\tC\tT\nrs6015519\t20\t58098827\tG\tA\n.\t20\t58114167\tT\tC\nrs1093454\t20\t59121456\tA\tG\nrs6010666\t20\t61381218\tG\tA\nrs4809452\t20\t61644330\tA\tG\nrs6090228\t20\t61706750\tC\tT\n"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-29T16:00:50.011303"
    },
    "Should run QC-only": {
        "content": [
            "::log:: Reference Panel Ranges: genome-wide\n::message:: Calculating QC Statistics\n::group type=message::\n<b>Statistics:</b>\nAlternative allele frequency > 0.5 sites: 2,296\nReference Overlap: 99.00 %\nMatch: 7,735\nAllele switch: 0\nStrand flip: 0\nStrand flip and allele switch: 0\nA/T, C/G genotypes: 0\n<b>Filtered sites:</b>\nFilter flag set: 0\nInvalid alleles: 0\nMultiallelic sites: 0\nDuplicated sites: 0\nNonSNP sites: 0\nMonomorphic sites: 11\nAllele mismatch: 0\nSNPs call rate < 90%: 0\n::endgroup::\n::group type=warning::\nExcluded sites in total: 11\nRemaining sites in total: 7,735\nSee snps-excluded.txt for details\nTyped only sites: 78\nSee typed-only.txt for details\n\n::endgroup::\n",
            "ID\tCHROM\tPOS\tREF\tALT\n.\t20\t161502\tC\tT\nSNP20-723376\t20\t775376\tT\tG\nrs535315\t20\t775842\tC\tT\n.\t20\t1438230\tC\tT\nrs6753\t20\t2638579\tT\tC\nrs1127354\t20\t3193842\tC\tA\nrs622099\t20\t3580168\tG\tA\nrs6052105\t20\t3811433\tT\tC\nrs6116189\t20\t4063862\tA\tG\nrs6139486\t20\t4578080\tG\tA\nSNP20-7489572\t20\t7541572\tA\tG\nrs6140334\t20\t7581461\tT\tG\n.\t20\t7928922\tA\tG\n.\t20\t8364799\tG\tT\n.\t20\t10243251\tT\tC\nrs1926936\t20\t11005238\tA\tC\nrs2103899\t20\t12173825\tT\tC\nrs6078983\t20\t13290738\tT\tC\n.\t20\t13582476\tA\tC\nrs4441543\t20\t13650883\tT\tG\nrs6042449\t20\t13929943\tC\tT\nrs2263684\t20\t13998098\tG\tA\nrs4490256\t20\t14794361\tA\tG\n.\t20\t14921606\tG\tA\nrs6075004\t20\t16066764\tG\tA\n.\t20\t16070941\tC\tT\n.\t20\t16116624\tG\tA\n.\t20\t16120763\tA\tG\nSNP20-17544155\t20\t17596155\tC\tA\nrs11960\t20\t17600357\tG\tA\nrs1073051\t20\t17845755\tT\tC\n.\t20\t18284404\tC\tT\n.\t20\t18389203\tT\tC\n.\t20\t19025636\tC\tT\n.\t20\t20130511\tT\tC\n.\t20\t20516020\tT\tC\n.\t20\t21628586\tC\tT\n.\t20\t21863397\tA\tG\n.\t20\t21939656\tC\tT\nrs6036488\t20\t23682810\tA\tG\n.\t20\t23725019\tA\tG\n.\t20\t24261606\tT\tC\n.\t20\t24691928\tT\tG\n.\t20\t24955981\tT\tC\n.\t20\t25136824\tA\tG\nrs6132909\t20\t25948327\tA\tG\n.\t20\t26307797\tT\tC\nrs1133358\t20\t31590686\tC\tA\nSNP20-35106470\t20\t35673056\tC\tT\n.\t20\t39048873\tA\tC\nrs2228246\t20\t39792063\tA\tG\nrs6072382\t20\t40093956\tA\tC\nrs3092716\t20\t40387411\tT\tC\nSNP20-40139709\t20\t40706295\tT\tC\nrs2425656\t20\t43506247\tG\tA\nrs1116100\t20\t43668473\tG\tA\n.\t20\t44011546\tC\tT\n.\t20\t46332216\tT\tC\nrs4810699\t20\t46435339\tT\tC\nrs4809715\t20\t47263751\tT\tC\n.\t20\t47479776\tC\tT\nrs522962\t20\t48143362\tT\tC\n.\t20\t48403371\tT\tC\nrs6067931\t20\t50395283\tT\tC\n.\t20\t50475418\tG\tA\n.\t20\t53251910\tG\tA\n.\t20\t53515863\tC\tA\n.\t20\t54264333\tC\tT\nrs6064338\t20\t54692704\tA\tG\nrs1047972\t20\t54961463\tT\tC\nrs4811958\t20\t56637477\tA\tG\n.\t20\t57304273\tC\tT\nrs6015519\t20\t58098827\tG\tA\n.\t20\t58114167\tT\tC\nrs1093454\t20\t59121456\tA\tG\nrs6010666\t20\t61381218\tG\tA\nrs4809452\t20\t61644330\tA\tG\nrs6090228\t20\t61706750\tC\tT\n",
            "ID\tCHROM\tPOS\tREF\tALT\tINFO\nrs6075795\t20\t2187430\tG\tA\tMonomorphic\n.\t20\t24404652\tC\tT\tMonomorphic\nrs6114728\t20\t24424220\tC\tT\tMonomorphic\nrs746748\t20\t25282967\tC\tT\tMonomorphic\n.\t20\t39005139\tC\tT\tMonomorphic\nrs2275801\t20\t45867852\tA\tG\tMonomorphic\n.\t20\t45873923\tC\tT\tMonomorphic\n.\t20\t48575315\tC\tT\tMonomorphic\nrs4811450\t20\t52269701\tC\tA\tMonomorphic\n.\t20\t54179856\tG\tT\tMonomorphic\nrs6027822\t20\t59363320\tG\tA\tMonomorphic\n"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-29T16:00:31.677665"
    }
}